webui/react/src/omnibar/Tree.ts

```diff
@@ -0,0 +1,105 @@
+
+import { StoreAction } from 'contexts/Store';
+import handleError, { ErrorType } from 'ErrorHandler';
+import { BaseNode, Children, getNodeChildren, isLeafNode,
+  isNLNode, isTreeNode, LeafNode, traverseTree, TreePath } from 'omnibar/AsyncTree';
+import { store } from 'omnibar/exposedStore';
+import root from 'omnibar/sampleTree';
+import { noOp } from 'services/utils';
+
+const SEPARATOR = ' ';
+
+interface TreeRequest {
+  path: TreePath;
+  query: string;
+}
+
+const parseInput = async (input: string): Promise<TreeRequest> => {
+  const sections = input.split(SEPARATOR);
+  const query = sections[sections.length-1];
+  const address = sections.slice(0,sections.length-1);
+  const path = await traverseTree(address, root);
+  return {
+    path,
+    query,
+  };
+};
+
+const absPathToAddress = (path: TreePath): string[] => (path.map(tn => tn.title).slice(1));
+
+const noResultsNode: LeafNode = {
+  closeBar: true,
+  label: 'no sub branches: exit',
+  onAction: noOp,
+  title: 'Exit',
+};
+
+const query = async (input: string): Promise<Children> => {
+  const { path, query } = await parseInput(input);
+  const node = path[path.length-1];
+  const children = await getNodeChildren(node);
+  const fuse = new Fuse(
+    children,
+    {
+      includeScore: false,
+      keys: [ 'title', 'aliases', 'label' ],
+      minMatchCharLength: 2,
+      shouldSort: true,
+      threshold: 0.4,
+    },
+  );
+  const matches = query === '' ? children : fuse.search(query).map(r => r.item);
+
+  if (isNLNode(node)) {
+    if (node.onCustomInput) {
+      const moreOptions = await node.onCustomInput(query);
+      matches.push(...moreOptions);
+    } else if (matches.length === 0) {
+      matches.push(noResultsNode);
+    }
+  }
+  return matches;
+};
+
+export const extension = async(input: string): Promise<Children> => {
+  try {
+    return await query(input);
+  } catch (e) {
+    handleError({
+      error: e,
+      message: 'failed to query omnibar',
+      type: ErrorType.Ui,
+    });
+    return [];
+  }
+};
+
+export const onAction = async (
+  item: BaseNode,
+  query: (input: string) => void,
+): Promise<void> => {
+  if (!!item && isTreeNode(item)) {
+    const input: HTMLInputElement|null = document.querySelector('#omnibar input[type="text"]');
+    if (!input) return Promise.resolve();
+    const { path } = await parseInput(input.value);
+    // update the omnibar text to reflect the current path
+    input.value = (path.length > 1 ? absPathToAddress(path).join(SEPARATOR) + SEPARATOR : '')
+        + item.title;
+    if (isLeafNode(item)) {
+      await item.onAction(item);
+      if (item.closeBar || path.find(n => n.title === 'goto')) {
+        store?.dispatch({ type: StoreAction.HideOmnibar });
+      } else {
+        message.info('Action executed.', 1);
+      }
+    } else {
+      // trigger the query.
+      input.value = input.value + SEPARATOR;
+      query(input.value);
+    }
+
+  }
+  return Promise.resolve();
+};
```

Comment:
**suggestion:** a bit of repeating, but these 3 things should probably just be in `omnibar/Components.tsx` directly, or they can be injected as a hook e.g. `useOmnibarTree` because they have React dependencies. I'm leaning more towards direct approach because `omnibar/Component` itself is tiny and there's no reusability concerns here.

Reply:
The rest can go inside an omnibar utility file.

Reply:
> Some observed quirks:
> 
> escape to dismiss the omnibar doesn't work for me (even right after triggering the omnibar with nothing typed)

should be fixed now. please check if cmd/ctrl + space works on mac to launch the bar

> UI suggestions:
 
> after user sees the help modal, when the user dismisses the modal (via okay button, escape key, enter key), re-focus on the omnibar and clear out the bar input, so the user can go right into using the omnibar again
> help modal to have direct links for each option that can then take the user back to the omnibar with the selected option partially filled out
> for help modal, for options that take in addition param, show the param. e.g. goto experiment <id> and kill notebook <id>

> break up combo words for commands. e.g. pauseExperiment to pause experiment

the only downside I see to this is that it'd be a bit slower to access as you can't get the full picture without getting the pause fully in. (that might be where tab to complete is helpful since enter to complete can risk executing something). This might also be addressed by allowing the user at each step to search through the fully resolved branches (same output as help) instead of just the immediate children (REF A)

> consistent command pattern of "action" followed by "noun" (or could be other formats, but what you have makes sense IMO). e.g. instead of dev serverAddress reset, reset serverAddress.

This makes sense. I mainly put it the other way to 1. better organize it: eg all serverAddress operations are under serverAddress 2. keep the tree narrower closer to the root 3. better discoverability: by looking at the nodes under dev you can't know what `dev reset` is but on the other hand looking at `dev serverAddress` you can guess operations relating to serverAddress can be found there.

> using Tab to select the highlighted option (after using the arrow keys) but NOT execute the option. Where as Enter would actually fire off the command if applicable. So if user typed hel and the highlighted option is help, hitting Enter would bring up the help modal, whereas Tab would just show help typed into the omnibar but not yet fired. So thinking of Tab as just an option selector, and Enter as both an option selector and also execution of the command.

> upon a match where there are options being shown, show the full command instead of just the next available word. So if I type goto , it would show goto **experiment**, goto **cluster**, etc (with the shared portion showing in lighter great, so the auto complete portion would be more emphasized)

doable. will need to make sure if the command gets the overflow is handled reasonably

> change "no sub branches: exit" to something like "no options matched". IMO, also thinking that it might be better to prevent Enter from dismissing the omnibar (only allowing escape key at this point)

Hmm do you think it probably serves our users better if it always stays around as opposed to sometime auto closing after a commnad? :thinking

> Nice to haves:
> 
> support for keyboard shortcuts for available options (added a screenshot below)



considering the suggestions, the work estimates, and impact estimates, I'd prioritize these 3 in this order to work on first.

- search against fully resolved branches
- merge mousehover and is selected
- break commands such as pauseExperiment

Reply:
> > Some observed quirks:
> > escape to dismiss the omnibar doesn't work for me (even right after triggering the omnibar with nothing typed)
> 
> should be fixed now. please check if cmd/ctrl + space works on mac to launch the bar
> 
works great!

> > UI suggestions:
> 
> > after user sees the help modal, when the user dismisses the modal (via okay button, escape key, enter key), re-focus on the omnibar and clear out the bar input, so the user can go right into using the omnibar again
> > help modal to have direct links for each option that can then take the user back to the omnibar with the selected option partially filled out
> > for help modal, for options that take in addition param, show the param. e.g. goto experiment  and kill notebook
> 
> > break up combo words for commands. e.g. pauseExperiment to pause experiment
> 
> the only downside I see to this is that it'd be a bit slower to access as you can't get the full picture without getting the pause fully in. (that might be where tab to complete is helpful since enter to complete can risk executing something). This might also be addressed by allowing the user at each step to search through the fully resolved branches (same output as help) instead of just the immediate children (REF A)
> 
After playing around with this more, I understand the limitations. The ideal thing would be to explore beyond just the one level deep and see more options that can potentially match. `pau` to be able to show options `pause experiement` and `pause trial` (made up). But that's probably not trivial to implement, but maybe we can provide that functionality as a future improvement?

> > consistent command pattern of "action" followed by "noun" (or could be other formats, but what you have makes sense IMO). e.g. instead of dev serverAddress reset, reset serverAddress.
> 
> This makes sense. I mainly put it the other way to 1. better organize it: eg all serverAddress operations are under serverAddress 2. keep the tree narrower closer to the root 3. better discoverability: by looking at the nodes under dev you can't know what `dev reset` is but on the other hand looking at `dev serverAddress` you can guess operations relating to serverAddress can be found there.
> 
I think this is related to the previous point, where if we can extend our look-ahead algorithm, we won't have to break a pattern to make it more useable.

> > using Tab to select the highlighted option (after using the arrow keys) but NOT execute the option. Where as Enter would actually fire off the command if applicable. So if user typed hel and the highlighted option is help, hitting Enter would bring up the help modal, whereas Tab would just show help typed into the omnibar but not yet fired. So thinking of Tab as just an option selector, and Enter as both an option selector and also execution of the command.
> 
> this is a missing feature from the omnibar and to add it I think would require a nontrivial change to the package.
> 
I see, yep not worth to do right now.

> > upon a match where there are options being shown, show the full command instead of just the next available word. So if I type goto , it would show goto **experiment**, goto **cluster**, etc (with the shared portion showing in lighter great, so the auto complete portion would be more emphasized)
> 
> doable but I'm not sure how much it'll add to the user experience. I think we could use the space better if we showed a description of what it would do instead of repeating what the user has just typed in the omnibar and can see right above the presented options.
> 
I think the full command match is more important to have when we are fuzzy matching across multiple node levels, and not just one level deep. If the user thinks of a command from a noun, it can better aid them. For example if they want to know what they can do with an experiment, they can type `exp` and the full commands `kill **exp**eriment <id>`, `pause **exp**eriment <id>`, `pause all **exp**eriments`, etc will show up. The motivation for this is to really make the omnibar more flexible and not force the user to think about what they want to do in a specific approach (which we have the WebUI to be there for).

> > change "no sub branches: exit" to something like "no options matched". IMO, also thinking that it might be better to prevent Enter from dismissing the omnibar (only allowing escape key at this point)
> 
> Hmm do you think it probably serves our users better if it always stays around as opposed to sometime auto closing after a commenad? :thinking . If escape wasn't so far away from the home row i'd feel better about not auto closing. maybe if I modify our hotkeys to allow closing with ctrl+space it'd help here
> 
This one is subjective and after looking at other omnibars, they generally always have a go-to / default option so there is not an `unmatched` option. So the way I'm thinking about this is that the user came to the omnibar for a reason. And if they go through a flow and ends up not finding a path, the question I asked myself is, would the user want to just end it there or try another path? I was leaning more towards the 2nd option. They also have the option click on any part of the overlay to dismiss in addition to the escape key.

> > Nice to haves:
> > support for keyboard shortcuts for available options (added a screenshot below)
> 
I saw the return icon thanks! 👍  we can place other shortcuts in the future!

> TODOs considering the suggestions, the work estimates, and impact estimates:
> 
> * [x]  merge mousehover and is selected
> * [x]  maybe remove auto closing
> * [ ]  search against fully resolved branches [could get expensive]
> * [ ]  maybe break commands such as pauseExperiment

Reply:
> I think the full command match is more important to have when we are fuzzy matching across multiple node levels,

yeah I agree we'll def want it for that

> I was leaning more towards the 2nd option. They also have the option click on any part of the overlay to dismiss in addition to the escape key.

I removed that.